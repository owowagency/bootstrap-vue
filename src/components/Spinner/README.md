# Spinner

```vue
<Spinner size="lg" />
```

## Imported Props

| Prop name | Description                  | Type                                             | Values | Default   |
| --------- | ---------------------------- | ------------------------------------------------ | ------ | --------- |
| size      | The size of the component    | [Size](../../composables/useSize) (string)       | -      | 'md'      |
| variant   | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'primary' |

## Props

| Prop name | Description             | Type | Values             | Default  |
| --------- | ----------------------- | ---- | ------------------ | -------- |
| type      | The type of the spinner | Type | `'border'\|'grow'` | 'border' |
