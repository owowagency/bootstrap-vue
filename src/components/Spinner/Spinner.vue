<template>
    <div :class="classes" />
</template>

<script lang="ts">
import {PropType, computed, toRef} from 'vue';
import useSize, {sizeProps} from '@/composables/useSize';
import useVariant, {variantProps} from '@/composables/useVariant';
import useClasses from '@/composables/useClasses';
import useStringTemplate from '@/composables/useStringTemplate';

export const types = ['border', 'grow'] as const;

type Type = typeof types[number];
</script>

<script lang="ts" setup>
const props = defineProps({
    ...sizeProps,
    /**
     * The type of the spinner
     * @values 'border'|'grow'
     */
    type: {
        type: String as PropType<Type>,
        default: 'border',
        validator: (t: Type) => types.includes(t),
    },
    ...variantProps,
});

const spinnerClass = useStringTemplate('spinner-{0}', toRef(props, 'type')).templatedString;

const {classes} = useClasses(computed(() => [
    useSize(props.size, `${spinnerClass.value}-{0}`).sizeClass.value,
    spinnerClass.value,
    useVariant(props.variant, 'text-{0}').variantClass.value,
]));
</script>

<docs>
```vue
<Spinner size="lg" />
```

## Imported Props

| Prop name | Description                  | Type                                             | Values | Default   |
| --------- | ---------------------------- | ------------------------------------------------ | ------ | --------- |
| size      | The size of the component    | [Size](../../composables/useSize) (string)       | -      | 'md'      |
| variant   | The variant of the component | [Variant](../../composables/useVariant) (string) | -      | 'primary' |
</docs>
