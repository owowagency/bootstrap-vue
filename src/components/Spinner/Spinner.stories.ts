import Spinner from '.';
import {sizes} from '@/composables/useSize';
import {types} from './Spinner.vue';
import {variants} from '@/composables/useVariant';

export default {
    title: 'Components/Spinner',
    component: Spinner,
    argTypes: {
        type: {
            control: {type: 'select'},
            options: types,
        },
        size: {
            control: {type: 'select'},
            options: sizes,
        },
        variant: {
            control: {type: 'select'},
            options: variants,
        },
    },
};

const template = (args) => ({
    components: {Spinner},
    setup: () => ({args}),
    template: '<Spinner v-bind="args" />',
});

export const Default = template.bind({});

export const Sizing = (args) => ({
    components: {Spinner},
    setup: () => ({args, sizes}),
    template: '<Spinner v-for="s of sizes" class="me-1" :size="s" v-bind="args" />',
});

export const Type = (args) => ({
    components: {Spinner},
    setup: () => ({args, types}),
    template: '<Spinner v-for="t of types" class="me-1" :type="t" v-bind="args" />',
});

export const Variants = (args) => ({
    components: {Spinner},
    setup: () => ({args, variants}),
    template: '<Spinner v-for="v of variants" class="me-1" :variant="v" v-bind="args" />',
});
