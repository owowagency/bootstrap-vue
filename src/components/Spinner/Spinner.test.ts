import Spinner from '.';
import {describe} from 'vitest';

describe('template', () => {
    componentRenderTest(Spinner);

    componentWrapperClassTest(Spinner, {type: 'grow'}, 'spinner-grow');

    componentWrapperClassTest(Spinner, {size: 'lg'}, 'spinner-border-lg');

    componentWrapperClassTest(Spinner, {variant: 'danger'}, 'text-danger');
});
