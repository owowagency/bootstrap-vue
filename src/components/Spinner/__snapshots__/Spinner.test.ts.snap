// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `"<div class=\\"spinner-border-md spinner-border text-primary\\"></div>"`;
